src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc

```diff
@@ -0,0 +1,174 @@
+    // based on timestamp
+
+    // get the response deque corresponding to the packet ID of the request deque
+    auto pos = resp_frames->find(packet_id);
+    // note that not finding a corresponding response deque is not indicative of error, as in
+    // case of MQTT packets that do not have responses like Publish with QOS 0
+    std::deque<mqtt::Message> empty_deque;
+    std::deque<mqtt::Message>& resp_deque = (pos != resp_frames->end()) ? pos->second : empty_deque;
+
+    // track the latest response timestamp to compare against request frame's timestamp later.
+    uint64_t latest_resp_ts = resp_deque.empty() ? 0 : resp_deque.back().timestamp_ns;
+    // finding the closest appropriate response from response deque in terms of timestamp and type
+    // for each request in the request deque
+    for (mqtt::Message& req_frame : req_deque) {
+      // getting the appropriate response match value for the request match key
+      uint8_t request_match_key =
+          (req_frame.control_packet_type << 4) | static_cast<uint8_t>(req_frame.dup) << 3 |
+          (req_frame.header_fields["qos"]) << 1 | static_cast<uint8_t>(req_frame.retain);
+      auto iter = MapRequestToResponse.find(request_match_key);
+      if (iter == MapRequestToResponse.end()) {
+        VLOG(1) << absl::Substitute("Could not find any responses for frame type = $0",
+                                    request_match_key);
+        continue;
+      }
+      if (iter->second == 0xff) {
+        // Request without responses found
+        req_frame.consumed = true;
+        latest_resp_ts = req_frame.timestamp_ns + 1;
+        mqtt::Message dummy_resp;
+        entries.push_back({std::move(req_frame), std::move(dummy_resp)});
+        continue;
+      }
+      uint8_t response_match_value = iter->second;
+
+      // finding the first response frame with timestamp greater than request frame
+      auto first_timestamp_iter =
+          std::lower_bound(resp_deque.begin(), resp_deque.end(), req_frame.timestamp_ns,
+                           [](const mqtt::Message& message, const uint64_t ts) {
+                             return ts > message.timestamp_ns;
+                           });
+      if (first_timestamp_iter == resp_deque.end()) {
+        VLOG(1) << absl::Substitute("Could not find any responses after timestamp = $0",
+                                    req_frame.timestamp_ns);
+        continue;
+      }
+
+      // finding the first appropriate response frame with the desired control packet type and flags
+      auto response_frame_iter = std::find_if(
+          first_timestamp_iter, resp_deque.end(), [response_match_value](mqtt::Message& message) {
+            return ((message.control_packet_type << 4) | (static_cast<uint8_t>(message.dup) << 3) |
+                    (message.header_fields["qos"] << 1) | static_cast<uint8_t>(message.retain)) ==
+                   response_match_value;
```

Comment:
This would be changed to this according to the function for this logic,
```
auto response_frame_iter = std::find_if(
          first_timestamp_iter, resp_deque.end(), [response_match_value](mqtt::Message& message) {
            return getMatchKey(message) == response_match_value;
          });
```

Reply:
Sounds good to me :+1: